lib/authorizeAccessToken.js

```diff
@@ -0,0 +1,216 @@
+/*!
+ * Copyright (c) 2021 Digital Bazaar, Inc. All rights reserved.
+ */
+import * as authorization from 'auth-header';
+import * as JWT from '@digitalbazaar/minimal-jwt';
+import base64url from 'base64url-universal';
+import {InvalidRequest, AccessDenied} from '@interop/oauth2-errors';
+import noopLogger from './noopLogger';
+
+const DEFAULT_ALLOWED_JWT_ALGS = new Set(['HS256', 'ES256']);
+
+/**
+ * Authorizes an incoming request for a route (parses Authorization: header,
+ * decodes access token, verifies it, checks the decoded token claims against
+ * the provided scope, sets the `req.clientMetadata` property).
+ *
+ * @param {object} options - Options hashmap.
+ * @param {string} options.scope - Scope required for the route guarded by this
+ *   `authorize()` usage.
+ * @param {Function} [options.customValidate] - Optional custom authorization
+ *   callback with the signature `async ({req, claims}) => {}`.
+ * @param {object} [options.cache] - Optional LRU cache to store the decoded
+ *   token claims). The cache API must be compatible with the lru-cache NPM
+ *   module.
+ * @param {string[]} options.validIssuers - List of allowed issuers of
+ *   access tokens (e.g. 'https://example.com').
+ * @param {Function} options.loadClientRegistration - A function that
+ *    loads a client.
+ * @param {object} options.tokenizer - Auto-rotating tokenizer.
+ * @param {object} [options.logger] - Logger.
+ * @returns {Function} Returns an Express.js middleware route handler.
+ */
+export function authorizeAccessToken({
+  scope, customValidate, cache, validIssuers, loadClientRegistration, tokenizer,
+  logger = noopLogger
+}) {
+  return async (req, res, next) => {
+    try {
+      // Extract access token from Authorization: header (throw otherwise)
+      let clientAccessToken;
+      try {
+        // parse throws if the authorization header is missing
+        const auth = authorization.parse(req.get('authorization'));
+        if(!(auth.scheme === 'Bearer' && auth.token)) {
+          throw new InvalidRequest({
+            description: 'Missing or invalid "authorization" header.'
+          });
+        }
+        ({token: clientAccessToken} = auth);
+      } catch(e) {
+        throw new InvalidRequest({
+          description: 'Missing or invalid "authorization" header.'
+        });
+      }
```

Comment:
This is a strange pattern because there errors thrown inside the try block are just dropped on the floor.  It appears that there are two possible sources of errors:
- `authorization.parse` could throw
-  The validation on L44 could fail and L45 could throw.

Here's what I think would be better.
- Move the validation on L44 out of this try catch and change THAT error message to 'Invalid "authorization" header.'
- Leave the parse on L43 and change the error message to be 'Missing "authorization" header.".

Reply:
Thanks. I have updated it here 37f62ff